File: tools/codec.py
"""息壤节点编解码器（v1.0）。

把「节点」（四属性）变成字节、再变回来。格式见 README。

字节序约定：所有多字节整数用**大端序**（网络字节序）。

一个节点的字节布局（无头部、无版本号）：

    [节点编号 16 字节] [父节点 16 字节] [节点名] [节点值]

- 节点名 = [1 字节长度] [UTF-8 字节]
- 节点值 = [类型标记 1 字节] [内容]
- 父节点 = nil UUID（16 字节全 0）表示「根节点」

值的大类（类型标记）与内容：

    0 空        无内容（纯容器节点，只有子节点）
    1 整数      8 字节（有符号 int64，大端）
    2 浮点数    8 字节（IEEE 754 double，大端）
    3 布尔      1 字节（0 = 假，1 = 真）
    4 文本      [4 字节长度] [UTF-8 字节]
    5 引用      16 字节 UUID（构成关联边）
    6 二进制块  [8 字节长度] [原始字节]

兼容性：旧标记（0-6）永不复用；要改就新增标记。
"""

import struct
import uuid
from dataclasses import dataclass
from typing import Optional, Tuple

# —— 大类标记（类型标记，1 字节）——
EMPTY = 0      # 空：纯容器节点，无值内容
INT = 1        # 整数
FLOAT = 2      # 浮点数
BOOL = 3       # 布尔
TEXT = 4       # 文本
REFERENCE = 5  # 引用：UUID，构成关联边
BLOB = 6       # 二进制块：内联原始字节

TAG_NAMES = {
    EMPTY: "空",
    INT: "整数",
    FLOAT: "浮点数",
    BOOL: "布尔",
    TEXT: "文本",
    REFERENCE: "引用",
    BLOB: "二进制块",
}

NIL_UUID = uuid.UUID(int=0)  # 全 0，代表「空」（根节点）


@dataclass
class Node:
    """息壤节点（四属性）。

    字段：
      id:     节点编号（UUID）
      parent: 父节点编号（UUID）；None 表示根节点
      name:   节点名（str）
      value:  节点值，形如 (tag, data)：
                (EMPTY, None)
                (INT, int)
                (FLOAT, float)
                (BOOL, bool)
                (TEXT, str)
                (REFERENCE, uuid.UUID)
                (BLOB, bytes)
    """
    id: uuid.UUID
    parent: Optional[uuid.UUID]
    name: str
    value: Tuple[int, object]


# ---------------------------------------------------------------------------
# 编码（节点 → 字节）
# ---------------------------------------------------------------------------

def _enc_short_text(s: str) -> bytes:
    """短文本（节点名）：1 字节长度 + UTF-8，上限 255 字节。"""
    raw = s.encode("utf-8")
    if len(raw) > 255:
        raise ValueError(f"文本超过 255 字节上限（{len(raw)} 字节）：{s!r}")
    return bytes([len(raw)]) + raw


def _enc_text(s: str) -> bytes:
    """文本值：4 字节长度 + UTF-8。"""
    raw = s.encode("utf-8")
    return len(raw).to_bytes(4, "big") + raw


def encode_value(tag: int, data: object) -> bytes:
    """把节点值编码成字节（含类型标记）。"""
    if tag == EMPTY:
        return bytes([EMPTY])
    if tag == INT:
        return bytes([INT]) + data.to_bytes(8, "big", signed=True)
    if tag == FLOAT:
        return bytes([FLOAT]) + struct.pack(">d", data)
    if tag == BOOL:
        return bytes([BOOL]) + (b"\x01" if data else b"\x00")
    if tag == TEXT:
        return bytes([TEXT]) + _enc_text(data)
    if tag == REFERENCE:
        return bytes([REFERENCE]) + data.bytes
    if tag == BLOB:
        return bytes([BLOB]) + len(data).to_bytes(8, "big") + data
    raise ValueError(f"未知类型标记：{tag}")


def encode_node(node: Node) -> bytes:
    """把节点编码成字节。"""
    pid = node.parent if node.parent is not None else NIL_UUID
    return (
        node.id.bytes
        + pid.bytes
        + _enc_short_text(node.name)
        + encode_value(node.value[0], node.value[1])
    )


# ---------------------------------------------------------------------------
# 解码（字节 → 节点）
# ---------------------------------------------------------------------------

def _dec_short_text(data: bytes, off: int) -> Tuple[str, int]:
    n = data[off]
    off += 1
    return data[off:off + n].decode("utf-8"), off + n


def _dec_text(data: bytes, off: int) -> Tuple[str, int]:
    n = int.from_bytes(data[off:off + 4], "big")
    off += 4
    return data[off:off + n].decode("utf-8"), off + n


def decode_value(data: bytes, off: int = 0) -> Tuple[Tuple[int, object], int]:
    """从字节解析节点值，返回 ((tag, data), 新偏移)。"""
    tag = data[off]
    off += 1
    if tag == EMPTY:
        v = None
    elif tag == INT:
        v = int.from_bytes(data[off:off + 8], "big", signed=True)
        off += 8
    elif tag == FLOAT:
        v = struct.unpack(">d", data[off:off + 8])[0]
        off += 8
    elif tag == BOOL:
        v = data[off] != 0
        off += 1
    elif tag == TEXT:
        v, off = _dec_text(data, off)
    elif tag == REFERENCE:
        v = uuid.UUID(bytes=data[off:off + 16])
        off += 16
    elif tag == BLOB:
        size = int.from_bytes(data[off:off + 8], "big")
        off += 8
        v = data[off:off + size]
        off += size
    else:
        raise ValueError(f"未知类型标记：{tag}")
    return (tag, v), off


def decode_node(data: bytes, off: int = 0) -> Tuple[Node, int]:
    """从字节解析节点，返回 (Node, 新偏移)。"""
    nid = uuid.UUID(bytes=data[off:off + 16])
    off += 16
    pid = uuid.UUID(bytes=data[off:off + 16])
    off += 16
    parent = None if pid == NIL_UUID else pid
    name, off = _dec_short_text(data, off)
    value, off = decode_value(data, off)
    return Node(id=nid, parent=parent, name=name, value=value), off

File: rust/core/src/codec.rs
//! 息壤节点编解码（内核）：节点 ↔ 字节。格式见 README，所有多字节整数为大端序。

use std::fmt;

/// 大类标记（类型标记，1 字节）。
pub const EMPTY: u8 = 0; // 空：纯容器节点，无值内容
pub const INT: u8 = 1; // 整数
pub const FLOAT: u8 = 2; // 浮点数
pub const BOOL: u8 = 3; // 布尔
pub const TEXT: u8 = 4; // 文本
pub const REFERENCE: u8 = 5; // 引用：UUID，构成关联边
pub const BLOB: u8 = 6; // 二进制块：内联原始字节

/// 16 字节 UUID（节点编号 / 引用目标）。
#[derive(Clone, Copy, PartialEq, Eq, Hash, Default)]
pub struct Uuid(pub [u8; 16]);

/// 全 0 UUID，代表「根节点」。
pub const NIL_UUID: Uuid = Uuid([0u8; 16]);

impl Uuid {
    pub fn is_nil(&self) -> bool {
        self.0 == [0u8; 16]
    }

    /// 解析 UUID 字符串（8-4-4-4-12，或去连字符的 32 位十六进制）；非法返回 None。
    pub fn parse(s: &str) -> Option<Uuid> {
        let hex: String = s.chars().filter(|c| *c != '-').collect();
        if hex.len() != 32 {
            return None;
        }
        let mut b = [0u8; 16];
        for (i, byte) in b.iter_mut().enumerate() {
            *byte = u8::from_str_radix(&hex[i * 2..i * 2 + 2], 16).ok()?;
        }
        Some(Uuid(b))
    }

    /// 生成随机 UUID v4（新增节点时发号）。
    pub fn random_v4() -> Uuid {
        Uuid(*uuid::Uuid::new_v4().as_bytes())
    }
}

impl fmt::Display for Uuid {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let b = &self.0;
        write!(
            f,
            "{:02x}{:02x}{:02x}{:02x}-{:02x}{:02x}-{:02x}{:02x}-{:02x}{:02x}-{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}",
            b[0], b[1], b[2], b[3], b[4], b[5], b[6], b[7], b[8], b[9], b[10], b[11], b[12], b[13], b[14], b[15]
        )
    }
}

impl fmt::Debug for Uuid {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Uuid({})", self)
    }
}

/// 节点值（7 大类）。
#[derive(Clone, PartialEq, Debug)]
pub enum Value {
    Empty,
    Int(i64),
    Float(f64),
    Bool(bool),
    Text(String),
    Reference(Uuid),
    Blob(Vec<u8>),
}

impl Value {
    pub fn tag(&self) -> u8 {
        match self {
            Value::Empty => EMPTY,
            Value::Int(_) => INT,
            Value::Float(_) => FLOAT,
            Value::Bool(_) => BOOL,
            Value::Text(_) => TEXT,
            Value::Reference(_) => REFERENCE,
            Value::Blob(_) => BLOB,
        }
    }
}

/// 节点（四属性：节点编号 → 父节点 → 节点名 → 节点值）。
#[derive(Clone, PartialEq, Debug)]
pub struct Node {
    pub id: Uuid,
    /// None = 根节点（父为全 0）。
    pub parent: Option<Uuid>,
    pub name: String,
    pub value: Value,
}

/// 解码错误。
#[derive(Debug, PartialEq)]
pub enum Error {
    /// 数据不足（文件截断）。
    Truncated,
    /// 类型标记非法。
    InvalidTag(u8),
    /// 名字 / 文本非法 UTF-8。
    InvalidUtf8,
    /// 名字超 255 字节。
    NameTooLong(usize),
}

// ---------------------------------------------------------------------------
// 编码
// ---------------------------------------------------------------------------

pub fn encode_value(v: &Value) -> Vec<u8> {
    let mut out = Vec::new();
    out.push(v.tag());
    match v {
        Value::Empty => {}
        Value::Int(n) => out.extend_from_slice(&n.to_be_bytes()),
        Value::Float(x) => out.extend_from_slice(&x.to_be_bytes()),
        Value::Bool(b) => out.push(if *b { 1 } else { 0 }),
        Value::Text(s) => {
            let b = s.as_bytes();
            out.extend_from_slice(&(b.len() as u32).to_be_bytes());
            out.extend_from_slice(b);
        }
        Value::Reference(u) => out.extend_from_slice(&u.0),
        Value::Blob(b) => {
            out.extend_from_slice(&(b.len() as u64).to_be_bytes());
            out.extend_from_slice(b);
        }
    }
    out
}

pub fn encode_node(n: &Node) -> Result<Vec<u8>, Error> {
    let nb = n.name.as_bytes();
    if nb.len() > 255 {
        return Err(Error::NameTooLong(nb.len()));
    }
    let mut out = Vec::with_capacity(32 + nb.len());
    out.extend_from_slice(&n.id.0);
    out.extend_from_slice(&n.parent.unwrap_or(NIL_UUID).0);
    out.push(nb.len() as u8);
    out.extend_from_slice(nb);
    out.extend_from_slice(&encode_value(&n.value));
    Ok(out)
}

// ---------------------------------------------------------------------------
// 解码
// ---------------------------------------------------------------------------

fn read_bytes<'a>(data: &'a [u8], off: &mut usize, n: usize) -> Result<&'a [u8], Error> {
    if data.len() < *off + n {
        return Err(Error::Truncated);
    }
    let s = &data[*off..*off + n];
    *off += n;
    Ok(s)
}

pub fn decode_value(data: &[u8], off: &mut usize) -> Result<Value, Error> {
    let tag = *read_bytes(data, off, 1)?.first().unwrap();
    match tag {
        EMPTY => Ok(Value::Empty),
        INT => {
            let b = read_bytes(data, off, 8)?;
            Ok(Value::Int(i64::from_be_bytes(b.try_into().unwrap())))
        }
        FLOAT => {
            let b = read_bytes(data, off, 8)?;
            Ok(Value::Float(f64::from_be_bytes(b.try_into().unwrap())))
        }
        BOOL => {
            let b = read_bytes(data, off, 1)?;
            Ok(Value::Bool(b[0] != 0))
        }
        TEXT => {
            let n = u32::from_be_bytes(read_bytes(data, off, 4)?.try_into().unwrap()) as usize;
            let b = read_bytes(data, off, n)?;
            Ok(Value::Text(
                String::from_utf8(b.to_vec()).map_err(|_| Error::InvalidUtf8)?,
            ))
        }
        REFERENCE => {
            let b = read_bytes(data, off, 16)?;
            Ok(Value::Reference(Uuid(b.try_into().unwrap())))
        }
        BLOB => {
            let n = u64::from_be_bytes(read_bytes(data, off, 8)?.try_into().unwrap()) as usize;
            let b = read_bytes(data, off, n)?;
            Ok(Value::Blob(b.to_vec()))
        }
        other => Err(Error::InvalidTag(other)),
    }
}

pub fn decode_node(data: &[u8], off: &mut usize) -> Result<Node, Error> {
    let id = Uuid(read_bytes(data, off, 16)?.try_into().unwrap());
    let p = Uuid(read_bytes(data, off, 16)?.try_into().unwrap());
    let parent = if p.is_nil() { None } else { Some(p) };
    let nlen = *read_bytes(data, off, 1)?.first().unwrap() as usize;
    let name = String::from_utf8(read_bytes(data, off, nlen)?.to_vec())
        .map_err(|_| Error::InvalidUtf8)?;
    let value = decode_value(data, off)?;
    Ok(Node {
        id,
        parent,
        name,
        value,
    })
}

/// 是否为「规范十进制整数」（无前导 0，如 0 / 51975 / -3；不含 01 / 007）。
pub fn is_canonical_int(s: &str) -> bool {
    let body = s.strip_prefix('-').unwrap_or(s);
    if body.is_empty() {
        return false;
    }
    if !body.bytes().all(|b| b.is_ascii_digit()) {
        return false;
    }
    body == "0" || !body.starts_with('0')
}

/// 命令行 / 界面输入 → 节点值：整数 / 浮点 / 布尔 / 文本。
/// 「01」「007」这类前导 0 字符串按**文本**存（否则会丢成整数 1 / 7）。
pub fn parse_value(s: &str) -> Value {
    if is_canonical_int(s) {
        if let Ok(i) = s.parse::<i64>() {
            return Value::Int(i);
        }
    }
    if let Ok(f) = s.parse::<f64>() {
        if s.contains('.') || s.contains('e') || s.contains('E') {
            return Value::Float(f);
        }
    }
    if s == "true" {
        return Value::Bool(true);
    }
    if s == "false" {
        return Value::Bool(false);
    }
    Value::Text(s.to_string())
}

/// 流式节点迭代器：逐个解码节点，不一次性加载全部（大文件 / 按需解析用）。
pub struct NodeIter<'a> {
    data: &'a [u8],
    off: usize,
}

impl<'a> NodeIter<'a> {
    pub fn new(data: &'a [u8]) -> Self {
        Self { data, off: 0 }
    }
}

impl<'a> Iterator for NodeIter<'a> {
    type Item = Result<Node, Error>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.off >= self.data.len() {
            return None;
        }
        Some(decode_node(self.data, &mut self.off))
    }
}

// ---------------------------------------------------------------------------
// 测试
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    fn node(id: [u8; 16], parent: Option<[u8; 16]>, name: &str, value: Value) -> Node {
        Node {
            id: Uuid(id),
            parent: parent.map(Uuid),
            name: name.into(),
            value,
        }
    }

    #[test]
    fn value_roundtrip() {
        let vals = vec![
            Value::Empty,
            Value::Int(2046),
            Value::Int(-42),
            Value::Float(1.5),
            Value::Bool(true),
            Value::Bool(false),
            Value::Text("灯".into()),
            Value::Text("hello world".into()),
            Value::Reference(Uuid([7u8; 16])),
            Value::Blob(vec![0xff, 0x00, 0x12]),
        ];
        for v in vals {
            let enc = encode_value(&v);
            let mut off = 0;
            let dec = decode_value(&enc, &mut off).unwrap();
            assert_eq!(v, dec);
            assert_eq!(off, enc.len());
        }
    }

    #[test]
    fn canonical_bytes() {
        // README 规范字节示例（对拍）
        assert_eq!(encode_value(&Value::Empty), vec![0x00]);
        assert_eq!(
            encode_value(&Value::Int(2046)),
            vec![0x01, 0, 0, 0, 0, 0, 0, 0x07, 0xfe]
        );
        assert_eq!(
            encode_value(&Value::Float(1.5)),
            vec![0x02, 0x3f, 0xf8, 0, 0, 0, 0, 0, 0]
        );
        assert_eq!(encode_value(&Value::Bool(true)), vec![0x03, 0x01]);
        assert_eq!(
            encode_value(&Value::Text("灯".into())),
            vec![0x04, 0, 0, 0, 3, 0xe7, 0x81, 0xaf]
        );
    }

    #[test]
    fn node_roundtrip() {
        let n = node(
            [1u8; 16],
            None,
            "灯",
            Value::Reference(Uuid([2u8; 16])),
        );
        let enc = encode_node(&n).unwrap();
        let mut off = 0;
        let dec = decode_node(&enc, &mut off).unwrap();
        assert_eq!(n, dec);
        assert_eq!(off, enc.len());
    }

    #[test]
    fn truncated() {
        let enc = encode_value(&Value::Text("灯".into()));
        let mut off = 0;
        assert_eq!(decode_value(&enc[..5], &mut off), Err(Error::Truncated));
    }

    #[test]
    fn invalid_tag() {
        let mut off = 0;
        assert_eq!(decode_value(&[99u8], &mut off), Err(Error::InvalidTag(99)));
    }

    #[test]
    fn name_too_long() {
        let n = node([1u8; 16], None, &"x".repeat(300), Value::Empty);
        assert!(matches!(encode_node(&n), Err(Error::NameTooLong(300))));
    }

    #[test]
    fn streaming_node_iter() {
        // 编码 3 个节点，流式迭代解码，数量一致且顺序正确
        let nodes = vec![
            node([1u8; 16], None, "a", Value::Empty),
            node([2u8; 16], Some([1u8; 16]), "b", Value::Int(42)),
            node([3u8; 16], Some([2u8; 16]), "c", Value::Text("灯".into())),
        ];
        let mut bytes = Vec::new();
        for n in &nodes {
            bytes.extend_from_slice(&encode_node(n).unwrap());
        }
        let iter = NodeIter::new(&bytes);
        let decoded: Vec<Node> = iter.map(|r| r.unwrap()).collect();
        assert_eq!(decoded.len(), 3);
        assert_eq!(decoded[1].name, "b");
        assert_eq!(decoded[2].value, Value::Text("灯".into()));
    }
}
